### ISSUE TITLE
Tensorflow V2 object detection onnx to tensorrt

### ISSUE THREAD

**Issue by Source82** [NVIDIA/TensorRT#1412]
Hello,

I have been trying to convert mobilenet v2 SSD following the suggestions given [here ](https://github.com/NVIDIA/TensorRT/issues/795) unfortunatley I am stuck with the error shown below. Basically I have tensorflow v2 model that has dynamic shapes, how do I go about making them static since one cannot freeze model on tensorflow v2.

# /usr/src/tensorrt/bin/trtexec --onnx=modele2-nms.onnx
[07/28/2021-10:36:44] [I] === Model Options ===
[07/28/2021-10:36:44] [I] Format: ONNX
[07/28/2021-10:36:44] [I] Model: modele2-nms.onnx
[07/28/2021-10:36:44] [I] Output:
[07/28/2021-10:36:44] [I] === Build Options ===
[07/28/2021-10:36:44] [I] Max batch: 1
[07/28/2021-10:36:44] [I] Workspace: 16 MB
[07/28/2021-10:36:44] [I] minTiming: 1
[07/28/2021-10:36:44] [I] avgTiming: 8
[07/28/2021-10:36:44] [I] Precision: FP32
[07/28/2021-10:36:44] [I] Calibration: 
[07/28/2021-10:36:44] [I] Safe mode: Disabled
[07/28/2021-10:36:44] [I] Save engine: 
[07/28/2021-10:36:44] [I] Load engine: 
[07/28/2021-10:36:44] [I] Builder Cache: Enabled
[07/28/2021-10:36:44] [I] NVTX verbosity: 0
[07/28/2021-10:36:44] [I] Inputs format: fp32:CHW
[07/28/2021-10:36:44] [I] Outputs format: fp32:CHW
[07/28/2021-10:36:44] [I] Input build shapes: model
[07/28/2021-10:36:44] [I] Input calibration shapes: model
[07/28/2021-10:36:44] [I] === System Options ===
[07/28/2021-10:36:44] [I] Device: 0
[07/28/2021-10:36:44] [I] DLACore: 
[07/28/2021-10:36:44] [I] Plugins:
[07/28/2021-10:36:44] [I] === Inference Options ===
[07/28/2021-10:36:44] [I] Batch: 1
[07/28/2021-10:36:44] [I] Input inference shapes: model
[07/28/2021-10:36:44] [I] Iterations: 10
[07/28/2021-10:36:44] [I] Duration: 3s (+ 200ms warm up)
[07/28/2021-10:36:44] [I] Sleep time: 0ms
[07/28/2021-10:36:44] [I] Streams: 1
[07/28/2021-10:36:44] [I] ExposeDMA: Disabled
[07/28/2021-10:36:44] [I] Spin-wait: Disabled
[07/28/2021-10:36:44] [I] Multithreading: Disabled
[07/28/2021-10:36:44] [I] CUDA Graph: Disabled
[07/28/2021-10:36:44] [I] Skip inference: Disabled
[07/28/2021-10:36:44] [I] Inputs:
[07/28/2021-10:36:44] [I] === Reporting Options ===
[07/28/2021-10:36:44] [I] Verbose: Disabled
[07/28/2021-10:36:44] [I] Averages: 10 inferences
[07/28/2021-10:36:44] [I] Percentile: 99
[07/28/2021-10:36:44] [I] Dump output: Disabled
[07/28/2021-10:36:44] [I] Profile: Disabled
[07/28/2021-10:36:44] [I] Export timing to JSON file: 
[07/28/2021-10:36:44] [I] Export output to JSON file: 
[07/28/2021-10:36:44] [I] Export profile to JSON file: 
[07/28/2021-10:36:44] [I] 
----------------------------------------------------------------
Input filename:   modele2-nms.onnx
ONNX IR version:  0.0.6
Opset version:    11
Producer name:    
Producer version: 
Domain:           
Model version:    0
Doc string:       
**----------------------------------------------------------------
[07/28/2021-10:36:48] [W] [TRT] onnx2trt_utils.cpp:220: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
[07/28/2021-10:36:48] [W] [TRT] onnx2trt_utils.cpp:246: One or more weights outside the range of INT32 was clamped
[07/28/2021-10:36:48] [I] [TRT] ModelImporter.cpp:135: No importer registered for op: BatchedNMS_TRT. Attempting to import as plugin.
[07/28/2021-10:36:48] [I] [TRT] builtin_op_importers.cpp:3659: Searching for plugin: BatchedNMS_TRT, plugin_version: 1, plugin_namespace: 
[07/28/2021-10:36:48] [I] [TRT] builtin_op_importers.cpp:3676: Successfully created plugin: BatchedNMS_TRT
[07/28/2021-10:36:48] [E] [TRT] onnx_graphsurgeon_node_0: PluginV2Layer must be V2DynamicExt when there are runtime input dimensions.
[07/28/2021-10:36:48] [E] [TRT] onnx_graphsurgeon_node_0: PluginV2Layer must be V2DynamicExt when there are runtime input dimensions.
[07/28/2021-10:36:48] [E] [TRT] onnx_graphsurgeon_node_0: PluginV2Layer must be V2DynamicExt when there are runtime input dimensions.
ERROR: builtin_op_importers.cpp:3040 In function importSlice:
[4] Assertion failed: -r <= axis && axis < r
[07/28/2021-10:36:48] [E] Failed to parse onnx file
[07/28/2021-10:36:48] [E] Parsing model failed
[07/28/2021-10:36:48] [E] Engine creation failed
[07/28/2021-10:36:48] [E] Engine set up failed
&&&& FAILED TensorRT.trtexec # /usr/src/tensorrt/bin/trtexec --onnx=modele2-nms.onnx**

**Comment by Source82** [comment]
@pranavm-nvidia please help with suggestions

**Comment by pranavm-nvidia** [comment]
You could try using Polygraphy to freeze the input shapes:
```bash
polygraphy surgeon sanitize modele2-nms.onnx \
    --override-input-shapes input_name:[1,3,224,224] \
    -o modele2-nms-static-shape.onnx
```

**Comment by Source82** [comment]
> You could try using Polygraphy to freeze the input shapes:
> 
> ```shell
> polygraphy surgeon sanitize modele2-nms.onnx \
>     --override-input-shapes input_name:[1,3,224,224] \
>     -o modele2-nms-static-shape.onnx
> ```

ok thanks

**Comment by Source82** [comment]
Hi @pranavm-nvidia 

Since I couldn't sort the issue with the model I mentioned earlier.

I have worked on the batchedNMS_TRT code using a different approach.

I trained the model using TF version 2 API, I then used tf2onnx to export two outputs (raw_detection_scores and raw_detection_boxes) since they are the two inputs into the BatchedNMS_TRT plugin. The model is attached (model2.onnx).

Then I used this code to append the NMS node, which has 4 outputs.

‘’’
take SSD onnx model that has two outputs (raw_detections, raw_detection_scores)
add BatchedNMS_TRT and 4 outputs
‘’’

import onnx_graphsurgeon as gs
import onnx
import numpy as np

input_model_path = “model2.onnx”

def append_nms(graph, num_classes, scoreThreshold, iouThreshold, keepTopK):
out_tensors = graph.outputs
boxes_input = graph.outputs[0]
scores_input = graph.outputs[1]
bs = out_tensors[0].shape[0]
print(bs)
num_priors = out_tensors[0].shape[1]
print(num_priors)

nms_attrs = {'shareLocation': True,
             'backgroundLabelId': -1,
             'numClasses': num_classes,
             'topK': num_priors,
             'keepTopK': keepTopK,
             'scoreThreshold': scoreThreshold,
             'iouThreshold': iouThreshold,
             'isNormalized': True,
             'clipBoxes': True}

num_detections = gs.Variable(name="num_detections", dtype=np.int32, shape=(bs, 1))
boxes =  gs.Variable(name="boxes", dtype=np.float32, shape=(bs, keepTopK, 4))
scores = gs.Variable(name="scores", dtype=np.float32, shape=(bs, keepTopK))
classes = gs.Variable(name="classes", dtype=np.float32, shape=(bs, keepTopK))

nms = gs.Node(op="BatchedNMS_TRT", attrs=nms_attrs, inputs=[boxes_input, scores_input], outputs=[num_detections, boxes, scores, classes])
graph.nodes.append(nms)
graph.outputs = [num_detections, boxes, scores, classes]

return graph
def add_nms_to_onnx(model_file, num_classes, confidenceThreshold=0.3, nmsThreshold=0.6, keepTopK=100, opset=11):
graph = gs.import_onnx(onnx.load(model_file))

graph = append_nms(graph, num_classes, confidenceThreshold, nmsThreshold, keepTopK)

# Remove unused nodes, and topologically sort the graph.
graph.cleanup().toposort().fold_constants().cleanup()

# Export the onnx graph from graphsurgeon
out_name = model_file[:-5]+'_nms.onnx'
onnx.save_model(gs.export_onnx(graph), out_name)

print("Saving the ONNX model to {}".format(out_name))
#convert and save
add_nms_to_onnx(input_model_path, 91, confidenceThreshold=0.3, nmsThreshold=0.6, keepTopK=100, opset=12)

After running trtexec everything seemed to work fine, the new model (i.e. model2_nms.onnx) was loaded and the BatchedNMS_TRT plugin was successfully created. However I have a sudden error (Aborted (core dumped)).

Please see attached logfile and the two onnx models. 
https://drive.google.com/file/d/15Hr3ebuXn3njuiVVEzWz4V28BWCUxpbq/view?usp=sharing

I will really appreciate your feedback.

Thanks

**Comment by azhurkevich** [comment]
@Source82 can you try a [sample](https://github.com/NVIDIA/TensorRT/tree/main/samples/python/tensorflow_object_detection_api) we've posted in fall. Please follow the [instructions](https://github.com/NVIDIA/TensorRT/blob/main/samples/python/tensorflow_object_detection_api/README.md) closely. This is for pretrained TFOD models from the [zoo](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_detection_zoo.md). If you have a custom model we cannot guarantee flawless conversion but it can certainly be done, if you are willing to read through code. I made sure there are a lot of comments that will help.

**Comment by nvpohanh** [comment]
This is because `BatchedDynamicNMS_TRT` instead of `BatchedNMS_TRT` should be used. @kevinch-nv @rajeevsrao Which one does ONNX parser use?

**Comment by rajeevsrao** [comment]
I think @Source82 explicitly added the `BatchedNMS_TRT` plugin using onnx-gs as a [workaround](https://github.com/NVIDIA/TensorRT/issues/795) for missing NMS support in the onnx parser prior to TRT 8.0.

Since TensorRT 8.x, our ONNX parser supports NonMaxSuppression` Op and [maps](https://github.com/onnx/onnx-tensorrt/blob/1da7332349d5b1196ccfa6dc719b839876f1e83e/builtin_op_importers.cpp#L2953) it to the new `EfficientNMS_ONNX_TRT` [plugin](https://github.com/NVIDIA/TensorRT/tree/main/plugin/efficientNMSPlugin).

@Source82 please refer to the instructions from @azhurkevich above regarding support for TFOD models.

**Comment by ttyio** [comment]
closing due to no activity for more than 3 weeks, please reopen if you still have question, thanks!
